### ISSUE TITLE
signInWithRedirect - social login using Google not working.

### ISSUE THREAD

**Issue by ashahjahan** [aws-amplify/amplify-js#13248]
### Before opening, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-js/issues?q=is%3Aissue+) and [discussions](https://github.com/aws-amplify/amplify-js/discussions).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### JavaScript Framework

React

### Amplify APIs

Authentication

### Amplify Version

v6

### Amplify Categories

auth

### Backend

Other

### Environment information

<details>

```
# Put output below this line

 System:
    OS: macOS 13.1
    CPU: (8) arm64 Apple M1
    Memory: 80.17 MB / 8.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 21.5.0 - ~/.nvm/versions/node/v21.5.0/bin/node
    npm: 10.3.0 - ~/.nvm/versions/node/v21.5.0/bin/npm
  Browsers:
    Chrome: 123.0.6312.123
    Safari: 16.2
  npmPackages:
    @mantine/core: ^7.6.1 => 7.6.1 
    @mantine/form: ^7.6.1 => 7.6.1 
    @mantine/hooks: ^7.6.1 => 7.6.1 
    @tabler/icons-react: ^2.47.0 => 2.47.0 
    @tanstack/react-router: ^1.19.1 => 1.19.1 
    @tanstack/router-devtools: ^1.19.1 => 1.19.1 
    @tanstack/router-vite-plugin: ^1.19.0 => 1.19.0 
    @types/react: ^18.2.56 => 18.2.61 
    @types/react-dom: ^18.2.19 => 18.2.19 
    @typescript-eslint/eslint-plugin: ^7.0.2 => 7.1.0 
    @typescript-eslint/parser: ^7.0.2 => 7.1.0 
    @vitejs/plugin-react: ^4.2.1 => 4.2.1 
    aws-amplify: ^6.0.21 => 6.0.21 
    aws-amplify/adapter-core:  undefined ()
    aws-amplify/analytics:  undefined ()
    aws-amplify/analytics/kinesis:  undefined ()
    aws-amplify/analytics/kinesis-firehose:  undefined ()
    aws-amplify/analytics/personalize:  undefined ()
    aws-amplify/analytics/pinpoint:  undefined ()
    aws-amplify/api:  undefined ()
    aws-amplify/api/server:  undefined ()
    aws-amplify/auth:  undefined ()
    aws-amplify/auth/cognito:  undefined ()
    aws-amplify/auth/cognito/server:  undefined ()
    aws-amplify/auth/enable-oauth-listener:  undefined ()
    aws-amplify/auth/server:  undefined ()
    aws-amplify/data:  undefined ()
    aws-amplify/data/server:  undefined ()
    aws-amplify/datastore:  undefined ()
    aws-amplify/in-app-messaging:  undefined ()
    aws-amplify/in-app-messaging/pinpoint:  undefined ()
    aws-amplify/push-notifications:  undefined ()
    aws-amplify/push-notifications/pinpoint:  undefined ()
    aws-amplify/storage:  undefined ()
    aws-amplify/storage/s3:  undefined ()
    aws-amplify/storage/s3/server:  undefined ()
    aws-amplify/storage/server:  undefined ()
    aws-amplify/utils:  undefined ()
    axios: ^1.6.7 => 1.6.7 
    eslint: ^8.56.0 => 8.57.0 
    eslint-plugin-react-hooks: ^4.6.0 => 4.6.0 
    eslint-plugin-react-refresh: ^0.4.5 => 0.4.5 
    postcss: ^8.4.35 => 8.4.35 
    postcss-preset-mantine: ^1.13.0 => 1.13.0 
    postcss-simple-vars: ^7.0.1 => 7.0.1 
    react: ^18.2.0 => 18.2.0 
    react-dom: ^18.2.0 => 18.2.0 
    react-query: ^3.39.3 => 3.39.3 
    typescript: ^5.2.2 => 5.3.3 
    vite: ^5.1.4 => 5.1.4 
  npmGlobalPackages:
    corepack: 0.23.0
    npm: 10.3.0



```

</details>

### Describe the bug

- I am using Cognito for auth, plugged into my React app on using vite  
- I am building my own UI flow and NOT using the hosted solution from Cognito  
- I am only using the Amplify auth library to interface with Cognito, I do not need access to the entire Amplify ecosystem, therefore:
- I am not using the Amplify CLI, but instead manually configuring using Amplify.configure({Auth: {...}})

When I use the line signInWithRedirect({ provider: "Google" }), I am redirected to an error page: https://xxx.us-east-1.amazoncognito.com/error?error=redirect_mismatch&client_id=xxx with a simple error page that states: "An error was encountered with the requested page."

The console gives the following errors:

1) GET https://xxx.amazoncognito.com/error?error=redirect_mismatch&client_id=xxx 400 (Bad Request)
openAuthSession @ localhost:5173/node_…js?v=44971122:10052
oauthSignIn @ localhost:5173/node_…js?v=44971122:10137
signInWithRedirect @ localhost:5173/node_…js?v=44971122:10107
await in signInWithRedirect (async)
onClick @ localhost:5173/src/r…t=1712839352616:156
callCallback2 @ localhost:5173/node_….js?v=44971122:3674
invokeGuardedCallbackDev @ localhost:5173/node_….js?v=44971122:3699
invokeGuardedCallback @ localhost:5173/node_….js?v=44971122:3733
invokeGuardedCallbackAndCatchFirstError @ localhost:5173/node_….js?v=44971122:3736
executeDispatch @ localhost:5173/node_….js?v=44971122:7016
processDispatchQueueItemsInOrder @ localhost:5173/node_….js?v=44971122:7036
processDispatchQueue @ localhost:5173/node_….js?v=44971122:7045
dispatchEventsForPlugins @ localhost:5173/node_….js?v=44971122:7053
(anonymous) @ localhost:5173/node_….js?v=44971122:7177
batchedUpdates$1 @ localhost:5173/node_…js?v=44971122:18909
batchedUpdates @ localhost:5173/node_….js?v=44971122:3579
dispatchEventForPluginEventSystem @ localhost:5173/node_….js?v=44971122:7176
dispatchEventWithEnableCapturePhaseSelectiveHydrationWithoutDiscreteEventReplay @ localhost:5173/node_….js?v=44971122:5478
dispatchEvent @ localhost:5173/node_….js?v=44971122:5472
dispatchDiscreteEvent @ localhost:5173/node_….js?v=44971122:5449

2) [Report Only] Refused to execute inline script because it violates the following Content Security Policy directive: "script-src https://d3oia8etllorh5.cloudfront.net https://xxx.auth.us-east-1.amazoncognito.com". Either the 'unsafe-inline' keyword, a hash ('sha256-iPbWFXQQ7k0RAW6EdruB+0wkXkyIRId3nYy5fwdyE/w='), or a nonce ('nonce-...') is required to enable inline execution.

I've searched for any sort of similar issue here, on stackoverflow and reddit.  The closest I've found is this individual who recently faced the same error #2 above: https://www.reddit.com/r/aws/comments/1ayy7hq/aws_cognito_giving_me_code_that_violates_their/

I've tried several different ways to fix this, but no luck so far.  Hoping for some insights here.

### Expected behavior

Redirect to a Google page for login.

### Reproduction steps

1. Install vite react app, AWS Amplify
2. Setup Google oauth key
3. Setup Cognito and integrate oauth keys + URLs
4. Configure AWS Amplify auth in code
5. Try to redirect to Google sign in using code

### Code Snippet

```javascript
// Put your code below this line.
Amplify.configure({
    Auth: {
      Cognito: {  
        userPoolClientId: 'xxx',
        userPoolId: 'xxx',
        loginWith: { // Optional
            username: true,
            email: true,
          oauth: {
            domain: 'xxx.us-east-1.amazoncognito.com',
            scopes: ["email", "openid", "profile"],
            redirectSignIn: ['localhost:5173/login'],
            redirectSignOut: ['localhost:5173/login'],
            responseType: 'code',
            providers: ["Google"]
          },
        }
      }
    }
  });

....

function handleSignInClick(customState: string) {
  signInWithRedirect({
    provider: "Google",
    customState
  });
}

```

### Log output

<details>

```
// Put your logs below this line


```

</details>

### Additional information and screenshots

<img width="846" alt="Screenshot 2024-04-12 at 1 34 06 PM" src="https://github.com/aws-amplify/amplify-js/assets/5751100/e7abf0a2-b31e-483b-8baa-c0be04ed5c56">

**Comment by ashahjahan** [comment]
Forgot to include this image of the errors in the console:

<img width="1382" alt="Screenshot 2024-04-12 at 2 00 59 PM" src="https://github.com/aws-amplify/amplify-js/assets/5751100/d49d5c7f-6e90-45b2-8ba1-ad055e2f21b8">

**Comment by cwomack** [comment]
Hello, @ashahjahan and sorry to hear you're running into this.  Couple of questions tied to your Auth setup to see if we can unblock you.  If you didn't set up your backend resources for Cognito with the CLI, are you trying to import existing resources? If so, can you confirm that the redirect URL's within your scoped Auth config match what is setup for your User Pool within the Cognito console?

**Comment by ashahjahan** [comment]
Hi @cwomack thanks for your response.  You are correct, I manually created the Cognito pool before looking at the Amplify solution.  I assumed I could just use the Amplify auth library to interface with the existing Cognito pool.

To answer your question, I just followed the guide located here: https://docs.amplify.aws/react/build-a-backend/auth/add-social-provider/ but I didn't see anything about adding a redirect URL within the Cognito UI.  Can you help me locate this within the Cognito UI as I am not using the CLI?  Thanks.

**Comment by ashahjahan** [comment]
Hi @cwomack based on your comment I was able to find the redirect URL in the Cognito panel under App Client -> Hosted UI.  I added the redirect and signout URLs there and it seems to have fixed the problem.  The misunderstanding was that I still needed to create a Hosted UI even though I believed I was building my own UI outside of the AWS ecosystem.

Although not needed for a Cognito login flow using Amplify auth libraries, it seems you still need to create a Hosted UI to handle the social login flows when using Amplify auth libraries.

Thanks for your help.

**Comment by adrianmtz20** [comment]
Hi @ashahjahan!
I have the same issue, i already add to google Oauth my pool, and custom my redirect url, in localhost my flow works perfect im able to signup with google but when i try using the amplify project with my own domain it shows me the same error as you at the beginning

¿any tip to solve it?

Thanks

**Comment by owlyowl** [comment]
I've had a very similar issue with it `404`ing but on the Google side.

The URL Amplify is generating is:
```
https://1057528857171-rc0mdtpn0cuvttuh95fe41vpjiflmkg8.apps.googleusercontent.com/oauth2/authorize?redirect_uri=http%3A%2F%2Flocalhost%3A3000%2F&response_type=code&client_id=348ovr4fk4pfq7slbbcg70c54u&identity_provider=Google&scope=openid%20email%20profile&state=dcopJW67GhpL2BaSdaFRlM0JTEBHtl5P&code_challenge=buEMB0Vj8mg7qtitgwSbbOWnWKRJu7H6w-qfZAf0HXU&code_challenge_method=S256
```
And that seems to throw up a 404. I've checked approved cognito domains and the like in Google's Cloud Console
![image](https://github.com/user-attachments/assets/32b7888e-a7af-48eb-9361-0d1b8868b90b)

And my cognito config seems correct:
```
export function createCognitoWithGoogleAuth(
    scope: Construct,
    props: CognitoWithGoogleAuthProps
) {
    const userPool = new UserPool(scope, `${props.appName}-userpool`, {
        userPoolName: `${props.appName}-userpool`,
        selfSignUpEnabled: true,
        accountRecovery: AccountRecovery.PHONE_AND_EMAIL,
        userVerification:{
            emailStyle: VerificationEmailStyle.CODE
        },
        autoVerify: {
            email: true
        },
        standardAttributes: {
            email: {
                required: true,
                mutable: true
            }
        }
    })

    const userPoolDomain = new UserPoolDomain(
        scope,
        `${props.appName}-userpooldomain`,
        {
            userPool,
            cognitoDomain: {
                domainPrefix: `${props.appName}`
            }
        }
    )

    const googleSecretValue = Secret.fromSecretNameV2(
        scope,
        'google.clientSecretName',
        `${props.appName}-googleoauth`
    )

    const googleProvider = new UserPoolIdentityProviderGoogle(
        scope,
        `${props.appName}-googleprovider`,
        {
            clientId: props.google.clientId,
            clientSecretValue: googleSecretValue.secretValue,
            scopes: ['openid', 'profile', 'email'],
            attributeMapping: {
                email: ProviderAttribute.GOOGLE_EMAIL,
                givenName: ProviderAttribute.GOOGLE_GIVEN_NAME,
                familyName: ProviderAttribute.GOOGLE_FAMILY_NAME,
                phoneNumber: ProviderAttribute.GOOGLE_PHONE_NUMBERS,
            },
            userPool,
        }
    )

    userPool.registerIdentityProvider(googleProvider)

    const userPoolClient = new UserPoolClient(
        scope,
        `${props.appName}-userpoolclient`,
        {
            userPool,
            oAuth: {
                flows: {
                    authorizationCodeGrant: true
                },
                callbackUrls: props.google.callbackUrls,
                logoutUrls: props.google.logoutUrls,
            }
        }
    )

    const identityPool = new IdentityPool(
        scope,
        `${props.appName}-identitypool`,
        {
            identityPoolName: `${props.appName}-identitypool`,
            allowUnauthenticatedIdentities: true,
            authenticationProviders: {
                userPools: [
                    new UserPoolAuthenticationProvider({
                        userPool: userPool,
                        userPoolClient: userPoolClient
                    })
                ]
            }
        }
    )

    return { userPool, userPoolClient, identityPool, userPoolDomain }
}
```

`signInWithRedirect({ provider: "Google" })`
